### PR TITLE
Embed browser

### LINKED ISSUE #44: Implement web view
- [ ] Embed JavaFX webview into GUI.
- [ ] Setup web engine structural skeleton

### PR DESCRIPTION
**Pull request by kxyoke**
Fixes #44.
@dorcastan you may want to use getCurrentPageUrl() to add bookmark.

### CODE REVIEW

File: src/main/java/seedu/mark/ui/BrowserPanel.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    /**
+     * Parses input into a valid url.
+     * If the input is simply lacking a protocol, http:// is prepended.
+     * Otherwise, the input is passed into google search.
+     * @param input non-url input
+     * @return a valid url based on given input.
+     */
+    private String parseNonUrl(String input) {
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: For future use; updated to pass codacy for now

File: src/main/java/seedu/mark/ui/BrowserPanel.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    private void loadDefaultPage() {
+        loadPage(DEFAULT_PAGE.toExternalForm());
+    }
+
+    /**
+     * Changes webview to site based on input entered in the address bar.
+     */
+    @FXML
+    private void handleAddressBarInput() {
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: Used in fxml; ignore codacy?

File: src/main/java/seedu/mark/ui/BrowserPanel.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    /**
+     * Initialises the address bar.
+     */
+    private void loadGuiAddress() {
+        addressBar.focusedProperty().addListener((observable, oldValue, newValue) -> {
+            Platform.runLater(() -> {
+                if (newValue) {
+                    logger.info("Address bar status: onfocus");
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: Actually I'm not very sure what we should/shouldn't log; but I guess it'll be easier to debug the GUI components.
I'll make the changes to the log message

File: src/main/java/seedu/mark/ui/BrowserPanel.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    ////////////////////// MAIN METHODS /////////////////////////
+
+    /**
+     * Loads page with url on the webview.
+     * @param url valid url
+     */
+    public void loadPage(String url) {
+        assert checkUrlValid(url) : "invalid url passed to webEngine";
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: I was thinking that whenever `loadPage()` is called, the string passed in will be valid (based on usage of the function: calling `checkUrlValid()` before `loadPage()` ). 
In that case should I do the check in `loadPage()` too?

File: src/main/java/seedu/mark/ui/BrowserPanel.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    private void loadDefaultPage() {
+        loadPage(DEFAULT_PAGE.toExternalForm());
+    }
+
+    /**
+     * Changes webview to site based on input entered in the address bar.
+     */
+    @FXML
+    private void handleAddressBarInput() {
+        //if address legit then load
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: Okay!

File: src/main/java/seedu/mark/ui/BrowserPanel.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+import seedu.mark.MainApp;
+import seedu.mark.commons.core.LogsCenter;
+
+/**
+ * The Browser Panel of Mark.
+ */
+public class BrowserPanel extends UiPart<Region> {
+
+    /** Default html page to be loaded when not connected to internet. */
+    public static final URL DEFAULT_PAGE =
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: Are you referring to the case where `DEFAULT_PAGE` is null?

File: src/main/java/seedu/mark/ui/BrowserPanel.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+            requireNonNull(MainApp.class
+                    .getResource(FXML_FILE_FOLDER + "default.html"));
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: Sounds good, I'll change the file name first, then when we have the cache I'll relocate the file.

File: src/main/java/seedu/mark/ui/BrowserPanel.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        //if is url without protocol, add protocol http://
+        //else google search it... how?
+        return input; //dummy code
+    }
+
+    /**
+     * Changes webview to google search page.
+     */
+    @FXML
+    private void handleGotoGoogle() {
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: I changed it to `gotoHomepage()` instead (feels more intuitive to me), don't know if that works.

File: src/main/resources/view/MainWindow.fxml [hunk 9]
```diff
@@ -45,12 +46,20 @@
...
-        </VBox>
+        <SplitPane fx:id="splitPane" dividerPositions="0.4" VBox.vgrow="ALWAYS">
+          <VBox fx:id="bookmarkList" styleClass="pane-with-border" minWidth="340" prefWidth="340" VBox.vgrow="ALWAYS">
+            <padding>
+              <Insets top="10" right="10" bottom="10" left="10" />
+            </padding>
+            <StackPane fx:id="bookmarkListPanelPlaceholder" VBox.vgrow="ALWAYS"/>
+          </VBox>
+
+          <StackPane fx:id="browserPlaceholder" prefWidth="340" >
```
[the remark this answers is not in the document]
Comment (kxyoke) [reply src/main/resources/view/MainWindow.fxml]: Ok, I'll settle this in issue #45 together with the other GUI tab view issues. 
Is it alright if I set the sidebar at a fixed width and the tab view as variable width?

File: src/main/java/seedu/mark/ui/BrowserPanel.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dorcastan) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: Ah ok! I missed the pre-condition in the javadocs.

File: src/main/java/seedu/mark/ui/BrowserPanel.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dorcastan) [reply src/main/java/seedu/mark/ui/BrowserPanel.java]: Oops, I meant the case when `MainApp.class.getResource(FXML_FILE_FOLDER + "default.html"))` is null. But I'm not very familiar with the use of `requireNonNull` in static variables, so if an error message is not useful then please ignore my comment!

File: src/main/resources/view/MainWindow.fxml (hunk 9 above)
[the remark this answers is not in the document]
Comment (dorcastan) [reply src/main/resources/view/MainWindow.fxml]: Sidebar meaning the bookmark list? Sounds good!

**Pull request merged**